ogr/services/github/auth_providers/abstract.py

```diff
@@ -0,0 +1,38 @@
+# Copyright Contributors to the Packit project.
+# SPDX-License-Identifier: MIT
+
+from typing import Optional
+
+import github
+
+
+class GithubAuthentication:
+    """
+    Represents a token manager for authentication via GitHubApp.
+    """
+
+    def get_pygithub_instance(self, namespace: str, repo: str) -> github.Github:
+        """
+        Returns GitHub instance for a requested repository,
+        authenticated using GitHub App
+
+        :param namespace: namespace of the repository
+        :param repo: name of the repository
+        :return: instance of github.Github
+        """
+        raise NotImplementedError()
+
+    @property
+    def pygithub_instance(self) -> "github.Github":
```

Comment:
This design (the authentication provider returning the instance of the service where it's being authenticated) is wired. Usually it's something like:

```
service = Service(AuthProvider())
```

and then in `Service.__init__()`:

```
token = auth_provider.get_token()
```

That is: the service receives an authentication provider which then is used to get the token.

Would it make sense to follow this pattern in this case, too?

Reply:
Two points that might answer your question, but TBH, I don't fully understand your question/proposal:

- We are instantiating the service when loading config. Sometimes they are not used => we want to ask for the token only when really needed.
- The token is per-project for GitHub apps and does not make sense to have one for the service.

Reply:
done

(can't be moved to `__init__` because of GitHub app tokens though)

Reply:
@lachmanfrantisek it removes the construction of `github.Github` from auth objects and constructs them in service with the provided token (which basically gets rid of the "same" `return` expression in all of them)

Reply:
> * The token is per-project for GitHub apps and does not make sense to have one for the service.

Apologies, I always forget that this is the case :|

> @lachmanfrantisek it removes the construction of github.Github from auth objects and constructs them in service with the provided token (which basically gets rid of the "same" return expression in all of them)

Exactly this was the design I was looking for, though failed to formulate it correctly. Thanks!